ministryofjustice/hmpps-authorization
# HAAR 1902 - Data layer

### What
- Data layer methods for the new UI
- Home page

### How to review
This is a large PR but shouldn't be too difficult once the architecture is understood. A quick explanations

#### Data layer

Our data layer architecture is all changes made in these directories
`/server/data`
`/server/interfaces`
`/server/mappers`
`/server/services`
`/server/testutils`
`/server/utils`

There are three classes of object

Model objects: `BaseClient`, `Client`, `ClientSecrets`
Response objects: `GetBaseClientResponse`, `ListBaseClientsResponse`, ...
Request objects: `AddBaseClientRequest`, `UpdateBaseClientRequest`, ...

Functions have been created in `/server/mappers` to map Model objects to Requests and Responses back to Model objects. The motivation for this is to allow our backend to evolve independently. When the API changes (and we are expecting change) we should be able to adjust our mapping functions and continue.

So the majority of added functionality is to take our backend API and for each restful method ...,

1. Add a method to an **api client** that takes a Request, does a network call, then returns a Response (see `/data/baseClientApiClient.ts`)
_Test - Use `nock` to mock the network - affirm that the baseClientApi method makes the correct network call and returns the right data_

2. Add **mapper** methods that convert model objects into requests and responses into model objects 
_Test - Not implemented - on hold until we are happy with the API stability_

3. Add a method to a **service** that takes a token and model objects as arguments. It will instantiate one or more api clients using the token then perform an operation using api methods. 
_Test - Mock api client methods - affirm that the method instantiates the necessary clients using the token - affirm it returns the mapped response back as a Model object_

Error handling happens at controller level so tests are happy path - for now at least

#### Home page
All file changes not in the above folders relate to making a homepage that demonstrates this approach is working, and fits with the MoJ template architecture
